neqo-http3/src/stream_type_reader.rs

```diff
@@ -6,57 +6,120 @@
-                ref mut reader,
-                stream_id,
-            } = self
-            {
+    fn read(&mut self, conn: &mut Connection) -> Res<(Option<u64>, bool)> {
+        match self {
+            NewStreamTypeReader::ReadType {
+                reader, stream_id, ..
+            }
+            | NewStreamTypeReader::ReadSessionId { reader, stream_id } => loop {
                 let to_read = reader.min_remaining();
                 let mut buf = vec![0; to_read];
-                match conn.stream_recv(*stream_id, &mut buf[..]) {
-                    Ok((0, false)) => {
-                        return None;
-                    }
-                    Ok((amount, false)) => {
-                        if let Some(res) = reader.consume(&mut Decoder::from(&buf[..amount])) {
-                            *self = NewStreamTypeReader::Done;
-                            return Some(res.into());
+                match conn.stream_recv(*stream_id, &mut buf[..])? {
+                    (0, f) => break Ok((None, f)),
+                    (amount, f) => {
+                        let res = reader.consume(&mut Decoder::from(&buf[..amount]));
+                        if res.is_some() || f {
+                            break Ok((res, f));
                         }
                     }
-                    Ok((_, true)) | Err(_) => {
+                }
+            },
+            _ => Ok((None, false)),
+        }
+    }
+
+    pub fn get_type(&mut self, conn: &mut Connection) -> Res<Option<NewStreamType>> {
+        loop {
+            let (output, fin) = self.read(conn)?;
+            if output.is_none() {
+                if fin {
+                    *self = NewStreamTypeReader::Done;
+                    break Err(Error::HttpStreamCreation);
+                }
+                break Ok(None);
+            }
+            let output = output.unwrap();
+            qtrace!("Decoded uint {}", output);
+            match self {
+                NewStreamTypeReader::ReadType {
+                    push_stream_allowed,
+                    stream_id,
+                    ..
+                } => {
+                    let res = NewStreamType::decode_stream_type(output, *push_stream_allowed);
+                    if res.is_err() {
+                        *self = NewStreamTypeReader::Done;
+                        return res;
+                    }
+                    let res = res.unwrap();
+                    if fin {
+                        *self = NewStreamTypeReader::Done;
+                        return self.map_stream_fin(res);
+                    }
+                    qtrace!("Decoded stream type {:?}", res);
+                    if res.is_some() {
                         *self = NewStreamTypeReader::Done;
-                        return None;
+                        break Ok(res);
+                    }
+                    // This is a WebTransportStream stream and it needs more data to be decoded.
```

Comment:
This isn't true yet.

Reply:
I did no remove all signs of WebTransport :)